oeps/oep-0051-bp-conventional-commits.rst

```diff
@@ -0,0 +1,181 @@
+
+* **style**: improve the styling of the code.
+
+* **test**: test-only changes. Adds missing tests or corrects existing tests. Tests accompanying other types of changes go into those commits.
+
+
+Breaking changes include an exclamation point as part of the type::
+
+    feat!: removed the ability to author courses
+
+Ideally, a single commit won't mix these types together.  If a commit does, use the most important type in the commit message.  The priority order for the types is generally: revert, feat, fix, perf, docs, test, build, refactor, style, chore.
+
+Scope
+-----
+
+The `Conventional Commits`_ spec includes a parenthesized scope after the type.  Open edX repos are large and varied, making standardization of scopes difficult.  We won't use scopes for now.  This could change in the future.
+
+Subject
+-------
+
+Commit message subjects should be short enough to fit on one line.  We aren't putting a hard limit on character length, but 70 characters is a good time to turn your attention to the body of the commit message.  Put more information in the body of the commit message to fully explain your change.  Jira or GitHub issue numbers can be included at the end of the subject.
+
+Body
+----
+
+The subject of the commit is rarely enough information to fully understand the commit.  The body can contain as much information as you like.  Be generous.  Take a moment to think about what you would want to know if someone else had authored this commit.
+
+The more information you can put in the body of the commit message, the better. It's not unreasonable to have two paragraphs of explanation in the body.
+
+Breaking changes especially should have detailed information about the implications and alternatives.
+
+
+Footer
+------
+
+We are not requiring any formalized footers in the commit message, though the use of  a ``BREAKING CHANGE:`` footer is encouraged as a way of providing more details about breaking changes.
+
+
+Discussion
+==========
+
+Conventional Commits asks us to categorize changes into a small number of categories.  There will naturally be cases where it isn't obvious which commit type to use.
+
+Choosing the commit type to use for a commit is important, but it's only the first step.  If you have doubts about which commit type to use, choose the highest-priority type that could apply.  Then write a detailed body explaining the full complexity of your change.
+
+**feat vs fix**: some user-visible changes to features could be classified as "feat" or "fix".  Choose "feat" if the change adds to the set of features.  Choose "fix" if the change affects how a feature behaves.  Yes, this is still subjective.  
+
+**breaking changes to features**: changing how a feature works is not a breaking change.  For example, users are sent to a new experience instead of the old experience. This is not a breaking change.  It should get a "feat" label, but not a "feat!" label.
+
+**DEPR**: removing deprecated code likely is a breaking change.
+
```

Comment:
This is an interesting point.  I think a lot of the existing conventional commit guidelines are aimed at automated versioning.  For that purpose, deprecation shouldn't be "feat!", because you don't want to bump the major version when you deprecate a feature.  You want to bump it when you remove the feature.

We've been focusing our discussion primarily on people reading the commits, and you are right, they need to sit up and notice when a deprecation happens.  One option would be a new label "depr:", but I also think the word "deprecate" will stand out:
```
feat: deprecate course creation
```
I'm in favor of labeling and signifiers, but there will also still be words. If we write them well, people will read them.

Reply:
As long as the word "deprecate" appears in the commit subject, when the code is deprecated _and_ when the code is removed, I think that is clear enough.

Reply:
Thanks for this discussion.

Can someone clarify for me where the discussion lands on what the commit message will be when a feature is actually removed?  Would it use the `feat!` prefix or the `fix!` prefix with the removing commit?

Reply:
The removing commit should use "feat!", since it is a breaking change (presumably) to a feature. "feat" should be used for any change in the feature set.